pkg/querier/tripperware/query_attribute_matcher.go

```diff
@@ -0,0 +1,218 @@
+				}
+			}
+		}
+
+		reqStats := stats.FromContext(r.Context())
+		reqStats.SetDataSelectMaxTime(maxTime)
+		reqStats.SetDataSelectMinTime(minTime)
+
+		if queryPriority := limits.QueryPriority(userStr); queryPriority.Enabled && len(queryPriority.Priorities) != 0 && query != "" {
+			for _, priority := range queryPriority.Priorities {
+				for _, attribute := range priority.QueryAttributes {
+					if matchAttributeForExpressionQuery(attribute, r, query, expr, now, minTime, maxTime) {
+						reqStats.SetPriority(priority.Priority)
+						return nil
+					}
+				}
+			}
+			reqStats.SetPriority(queryPriority.DefaultPriority)
+		}
+	}
+
+	if queryReject := limits.QueryRejection(userStr); queryReject.Enabled && isMetadataQuery {
+		for _, attribute := range queryReject.QueryAttributes {
+
+			if matchAttributeForMetadataQuery(attribute, r, now) {
+				rejectedQueriesPerTenant.WithLabelValues(userStr).Inc()
+				return httpgrpc.Errorf(http.StatusUnprocessableEntity, queryRejectErrorMessage)
+			}
+		}
+	}
+
+	return nil
+}
+
+func matchAttributeForExpressionQuery(attribute validation.QueryAttribute, r *http.Request, query string, expr parser.Expr, now time.Time, minTime, maxTime int64) bool {
+	if attribute.Regex != "" && attribute.Regex != ".*" && attribute.Regex != ".+" {
+		if attribute.CompiledRegex != nil && !attribute.CompiledRegex.MatchString(query) {
+			return false
+		}
+	}
+
+	if !isWithinTimeAttributes(attribute.TimeWindow, now, minTime, maxTime) {
+		return false
+	}
+
+	if !isWithinTimeRangeAttribute(attribute.TimeRangeLimit, maxTime-minTime) {
+		return false
+	}
+
+	if !isWithinQueryStepLimit(attribute.QueryStepLimit, r, expr) {
+		return false
+	}
+
+	if attribute.UserAgent != "" && attribute.UserAgent != r.Header.Get("User-Agent") {
+		return false
+	}
+
+	if attribute.DashboardUID != "" && attribute.DashboardUID != r.Header.Get("X-Dashboard-Uid") {
+		return false
+	}
+
+	if attribute.PanelID != "" && attribute.PanelID != r.Header.Get("X-Panel-Id") {
+		return false
+	}
+
+	return true
+}
+
+func matchAttributeForMetadataQuery(attribute validation.QueryAttribute, r *http.Request, now time.Time) bool {
+	if attribute.Regex != "" && attribute.Regex != ".*" && attribute.Regex != ".+" && attribute.CompiledRegex != nil {
+		for _, matcher := range r.Form["match[]"] {
+			if attribute.CompiledRegex.MatchString(matcher) {
+				continue
+			}
+			return false
+		}
+	}
+
+	startTime, _ := util.ParseTime(r.FormValue("start"))
+	endTime, _ := util.ParseTime(r.FormValue("end"))
+
+	if !isWithinTimeAttributes(attribute.TimeWindow, now, startTime, endTime) {
+		return false
+	}
+
+	if startTime != 0 && endTime != 0 && !isWithinTimeRangeAttribute(attribute.TimeRangeLimit, endTime-startTime) {
+		return false
+	}
+
+	if attribute.UserAgent != "" && attribute.UserAgent != r.Header.Get("User-Agent") {
+		return false
+	}
+
+	if attribute.DashboardUID != "" && attribute.DashboardUID != r.Header.Get("X-Dashboard-Uid") {
```

Comment:
Are we sure metadata queries from Grafana have dashboard IDs and Panel IDs? It would be better to validate that

Reply:
They don't have, but I remember there was expectation them to be added in metadata queries as well. Currently, we can just skip this attributes on query rejections configs for metadata queries and it will ignore it. And in the future if they will be provided, we can start using it right away.

Reply:
There is no timeline for it AFAIK. I will remove this for now, and it can be easily added after grafana starts sending those headers for metadata queries.

Reply:
removed